server/src/main/java/org/elasticsearch/index/fielddata/SortedBinaryDocValues.java

```diff
@@ -42,4 +42,55 @@ public abstract class SortedBinaryDocValues {
      */
     public abstract BytesRef nextValue() throws IOException;
 
+    /**
+     * Indicates the sparsity of the values for this field.
+     */
+    public Sparsity getSparsity() {
```

Comment:
This is the main change. Optionally `SortedBinaryDocValues` can now indicate the value mode and sparsity. The only classes that implement this are `SeparateCounts` (via doc value skipper on count field) and `SingletonSortedBinaryDocValues` (is by definition single valued). 

Then `SingleValueMatchQuery` uses this new capability of `SortedBinaryDocValues` to be more optimal when value mode is single value and rewrites to match all docs if sparsity is dense and value mode is single value.